### ISSUE TITLE
AI unable to control mechs when mech is not on the same z-level as the AI core.

### ISSUE THREAD

**Issue by TheMidnightRose** [tgstation/tgstation#45691]
Issue reported from  Round ID: 114967 (/tg/Station Bagil [ENGLISH] [US-WEST] [100% FREE LAG])

Reporting client version: 512.1472

Reproduction of this bug is simple, Be AI, get any mech with equipment and go to lavaland via the Remote Beacon, die when you are unable to fight anything while trying to rescue miners.

**Comment by PKPenguin321** [comment]
Not sure that this is actually a bug since I believe we intentionally restrict AIs to the z-level their core is on? Does hijacking a mech fully shunt you into it or is it just remote control?

**Comment by Anonmare** [comment]
Remote control is kind of weird since it puts your mind onside of the mech IIRC. It would be inconsistent if it were intended as AI shells can operate off native z level without issue.

**Comment by GunHog** [comment]
That is extra weird. Lore-wise, I want to say that AIs control mechs by uploading themselves into the mech, so the mech serves as the temporary core. A virus infecting your machine would not disappear just because you left the network. As for shells, AIs control shells like an RC car, via the binary comms link that all AIs and borgs share. This is not limited by Z-level.

Gameplay-wise, Z-level restrictions were added well after AI mechs and shells. I do not know the intent of the person who coded that, but I personally never intended Z-level limitations for either mechs nor shells. At most, an AI could be denied access to mechs outside the Z-level, but *once inside*, they should operate normally, because they are physically located inside the mech.

Adding a Z-level restriciton to shells does not make sense either, because all other Borg/AI interactions work regardless of location.

**Comment by PKPenguin321** [comment]
Going to call it a bug then, AI probably shouldn't get softlocked when a *temporary* mech core leaves the station. It should be forcibly disconnected instead.

**Comment by zxaber** [comment]
Yeah, AI code is fucky and mech code is fucky, so together they're double fucky.

When an AI "remote controls" a mech, the whole AI mob is actually just directly moved into the mech, and it leaves behind an empty core object. So technically the AI is on whatever Z level the mech is.

**Comment by nemvar** [comment]
Whoever unfucks both mech and AI code gets a cookie from me.

**Comment by GunHog** [comment]
>When an AI "remote controls" a mech, the whole AI mob is actually just directly moved into the mech, and it leaves behind an empty core object. So technically the AI is on whatever Z level the mech is.

Correct. The mech serves as your temporary core, similar to the shunt mechanic. Your original core can be destroyed without it affecting you until you try to return to it - in which case it is now your permanent core until someone puts you in a new one via carding.

**Comment by timothymtorres** [comment]
This affects posibrains as well apparently.

**Comment by Sealed101** [comment]
cannot reproduce this with posibrain mechs - the verbs don't do anything; however being unable to move z-levels in a mech in no gravity is quite bad

ais can control mechs on off-station z-levels but also can't use the verbs to move the mechs up or down

**Comment by Sealed101** [comment]
not sure what i was on when i brought up verbs into this i think there was some other issue report for that
either way can't repro, ais mechs work fine on other zlevels. my guess is it was disconnecting the AI because moving into an area with no camera coverage is supposed to disconnect the AI, and changing zlevel could place the mech into technically static. guess which part of that is broken (and honestly needs some buffer time before breaking the connection)
